plugins/tedge_apama_plugin/src/main.rs

```diff
@@ -50,123 +49,217 @@ const TEDGE_APAMA_PROJECT_DIR: &str = "/etc/tedge/apama/project";
-        } => {
-            let archive_path = Path::new(&file_path);
-            let archive_file = File::open(&archive_path)?;
+        } => match apama_module_from_string(&module)? {
+            ApamaModule::Project(_) => install_project(Path::new(&file_path)),
+            ApamaModule::MonFile(monitor_name) => {
+                install_or_update_monitor(&monitor_name, &file_path)
+            }
+        },
 
-            let mut archive = zip::ZipArchive::new(archive_file)?;
+        PluginOp::Remove { module, version: _ } => match apama_module_from_string(&module)? {
+            ApamaModule::Project(project_name) => remove_project(&project_name),
+            ApamaModule::MonFile(monitor_name) => remove_monitor(&monitor_name),
+        },
+    }
+}
 
-            // TODO: Validate the zip to be valid apama project before extraction?
-            println!("Extracting the archive at {}", file_path);
-            archive.extract(tmp_apama_project_path)?;
-            println!("Extraction successful");
+fn apama_module_from_string(module: &str) -> Result<ApamaModule, InternalError> {
+    match module.rsplit_once("::") {
+        Some((prefix, suffix)) => match suffix {
+            PROJECT_SUFFIX => Ok(ApamaModule::Project(prefix.into())),
+            MON_SUFFIX => Ok(ApamaModule::MonFile(prefix.into())),
+            unsupported_type => Err(InternalError::UnsupportedModuleType {
+                module_type: unsupported_type.into(),
+            }),
+        },
+        None => Err(InternalError::ModuleTypeNotProvided {
+            module_name: module.into(),
+        }),
+    }
+}
 
-            // Deleting existing project as the rename API expects the target dir to be empty
-            if tedge_apama_project_path.exists() {
-                println!("Removing existing project at {}", TEDGE_APAMA_PROJECT_DIR);
-                fs::remove_dir_all(tedge_apama_project_path)?;
-                println!("Removal of existing project successful");
-            }
+fn install_project(project_archive_path: &Path) -> Result<(), InternalError> {
+    let tedge_apama_project_path: &Path = Path::new(TEDGE_APAMA_PROJECT_DIR);
+    let tmp_apama_project_path: &Path = Path::new(TMP_APAMA_PROJECT_DIR);
 
-            println!(
-                "Installing newly extracted project to {}",
-                TEDGE_APAMA_PROJECT_DIR
-            );
-            fs::create_dir_all(tedge_apama_project_path)?;
-            fs::rename(tmp_apama_project_path, tedge_apama_project_path)?;
-            println!("Installation of new project successful");
+    let archive_file = File::open(&project_archive_path)?;
 
-            println!("Restarting apama to load the new project");
-            run_cmd("service", "apama restart")?;
-            println!("Restart of apama service successful");
+    let mut archive = zip::ZipArchive::new(archive_file)?;
 
-            success
-        }
+    // TODO: Validate the zip to be valid apama project before extraction?
```

Comment:
Good to validate.

Reply:
Does Apama provides some kind of `check` command? If not I would do nothing since *only* Apama can tell for sure that everything is fine. If an improvement is required I would be to rollback in case  `apama restart` fails.

Reply:
Apama doesn't provide any strict validations like that. They would even accept and empty directory as a "project". The only requirement is that the project must be in a directory named `project`. Currently the zip file is expected to have the contents of the project directly at the top level itself and we unzip and move these contents to the `/tedge/apama/project` directory. I could enforce a stricter structure to the zip file to have that `project` directory at the top level and all the project files under that directory which we don't validate. We'll just validate if the zip contains a top level directory named `project` fail if that's not the case.